src/game/connect4.rs

```diff
@@ -65,20 +84,224 @@ impl GameAdapter for Connect4Adapter<'_> {
             return Err(actix_web::Error::from(GameAdapterError::InvalidGameState(
                 self.state,
             )));
+        }
+        let request_payload = serde_json::from_value::<Connect4RequestPayload>(game_move.payload)?;
+        let column = request_payload.column;
+        let player = match self.game.turn {
+            Token::Red => self.players.get(0).unwrap().clone(),
+            Token::Blue => self.players.get(1).unwrap().clone(),
         };
-        todo!()
+        let user = game_move.player;
+
+        if player != user {
+            return Err(actix_web::Error::from(
+                GameAdapterError::WrongPlayerRequest(user),
+            )); // return the one who made the request
+        }
+        self.game.moves(column)?;
+        let win = self.game.winning_move(column);
+        let draw = self.game.is_game_drawn();
+        if win {
+            let winner = match self.game.turn {
+                Token::Red => self.players.get(0).unwrap().clone(),
+                Token::Blue => self.players.get(1).unwrap().clone(),
+            };
+            self.winner.push(winner);
+        }
+        if win || draw {
+            self.game.completed = true;
+        } else {
+            self.game.switch_token();
+        }
+        Ok(())
     }
 
     fn get_encoded_state(&self) -> actix_web::Result<GenericGameState> {
+        let encoded_board = self
+            .game
+            .board
+            .iter()
+            .map(|col| {
+                col.iter()
+                    .map(|&token| match token {
+                        Token::Red => self.players.get(0),
+                        Token::Blue => self.players.get(1),
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
         Ok(GenericGameState {
             game: "connect_4".to_string(),
             players: self.players.clone(),
             state: self.state,
             can_move: vec![self.next_move.clone()],
```

Comment:
Instead of having `self.next_move` in your code, just convert `turn` into the first or second player depending on if it is red or blue (unless the game hasn't started yet or it has ended, in which case just have `can_move` be an empty vector)